src/lib.ts

```diff
@@ -0,0 +1,320 @@
+	'X-Amz-Content-Sha256'?: string;
+	'X-Amz-Credential'?: string;
+	'X-Amz-SignedHeaders'?: string;
+	'X-Amz-Expires'?: string;
+	'X-Amz-Signature'?: string;
+	'X-Amz-Date'?: string;
+	'X-Amz-Security-Token'?: string;
+}
+
+function getDimArray(dims: string | number[], zoom: number = 1): (number | null)[] {
+	var dimArr = typeof dims === 'string' ? dims.split(',') : dims;
+	return dimArr.map(function (v) {
+		return Math.round(Number(v) * zoom) || null;
+	});
+}
+
+function clamp(val: number | string, min: number, max: number): number {
+	return Math.min(Math.max(Number(val), min), max);
+}
+
+export async function getS3File(
+	config: S3ClientConfig & { bucket: string },
+	key: string,
+	args: Args
+): Promise<GetObjectCommandOutput> {
+	const s3 = new S3Client({
+		...config,
+		signer: {
+			sign: async (request) => {
+				const isPresigned = !!args['X-Amz-Algorithm'];
+				if (!isPresigned) {
+					return request;
+				}
+				const presignedParamNames = [
+					'X-Amz-Algorithm',
+					'X-Amz-Content-Sha256',
+					'X-Amz-Credential',
+					'X-Amz-SignedHeaders',
+					'X-Amz-Expires',
+					'X-Amz-Signature',
+					'X-Amz-Date',
+					'X-Amz-Security-Token',
+				] as const;
+				const presignedParams: { [K in (typeof presignedParamNames)[number]]?: string } = {};
+				const signedHeaders = args['X-Amz-SignedHeaders']?.split(';') || [];
+
+				for (const paramName of presignedParamNames) {
+					if (args[paramName]) {
+						presignedParams[paramName] = args[paramName];
+					}
+				}
+
+				const headers: typeof request.headers = {};
+				for (const header in request.headers) {
+					if (signedHeaders.includes(header.toLowerCase())) {
+						headers[header] = request.headers[header];
+					}
+				}
+				request.query = presignedParams;
+
+				request.headers = headers;
+				return request;
+			},
+		},
+	});
+
+	const command = new GetObjectCommand({
+		Bucket: config.bucket,
+		Key: key,
+	});
+
+	return s3.send(command);
+}
+
+// return a default compression value based on a logarithmic scale
+// defaultValue = 100, zoom = 2; = 65
+// defaultValue = 80, zoom = 2; = 50
+// defaultValue = 100, zoom = 1.5; = 86
+// defaultValue = 80, zoom = 1.5; = 68
+function applyZoomCompression(defaultValue: number, zoom: number): number {
+	const value = Math.round(defaultValue - (Math.log(zoom) / Math.log(defaultValue / zoom)) * (defaultValue * zoom));
+	const min = Math.round(defaultValue / zoom);
+	return clamp(value, min, defaultValue);
+}
+
+export async function resizeBuffer(
+	buffer: Buffer | Uint8Array,
+	args: Args
+): Promise<{ data: Buffer; info: sharp.OutputInfo & { errors: string } }> {
+	const image = sharp(buffer as Buffer, { failOnError: false, animated: true }).withMetadata();
+
+	// check we can get valid metadata
+	const metadata = await image.metadata();
+
+	// auto rotate based on orientation EXIF data.
+	image.rotate();
+
+	// validate args, remove from the object if not valid
+	var errors: string[] = [];
```

Comment:
Set to const; push is not mutating the var, it's mutating the value

```suggestion
	const errors: string[] = [];
```